ipa-core/src/protocol/basics/share_validation.rs

```diff
@@ -0,0 +1,137 @@
+use futures_util::future::try_join;
+
+use crate::{
+    error::Error,
+    helpers::{Direction, MpcReceivingEnd, SendingEnd},
+    protocol::{
+        context::Context,
+        ipa_prf::malicious_security::hashing::{compute_hash, Hash},
+        RecordId,
+    },
+    secret_sharing::SharedValue,
+};
+
+/// This function checks that a vector of shares are consistent across helpers
+/// i.e. `H1` holds `(x0,x1)`, `H2` holds `(x1,x2)`, `H3` holds `(x2,x0)`
+/// the function verifies that `H1.x0 == H3.x0`, `H1.x1 == H2.x1`, `H2.x2 == H3.x2`
+///
+/// We use a hash based approach that is secure in the random oracle model
+/// further, only one of left and right helper check that it is zero
+/// this is might not be sufficient in some applications to prevent malicious behavior
+///
+/// The left helper simply hashes the vector and sends it to the right,
+/// the right helper negates his vector, hashes it and compares it to the received hash
+///
+/// # Errors
+/// propagates errors from send and receive
+pub async fn validate_replicated_shares<C, S>(
+    ctx: C,
+    input_left: &[S],
```

Comment:
is there a reason why you're not taking `Replicated<S>` instead?

Reply:
Yes. This check is for the verification of the proofs. I will have 3 provers and 6 verifiers. The two verifiers for each prover need to perform a zero check. You could considers this a 2 party computation where we have 2 out of 2 non-replicated shares and they want to check whether the shares sum up to zero. 

We could interpret the shares as replicated secret shares. One issue might be that the shares won't be consistent so if we add an automated consistency check for replicated shares, it would fail here.

Reply:
You're right that https://github.com/private-attribution/ipa/pull/936 is quite different. I am not sure whether we should try to make it compatible. In my use-case, we really don't need a stream since I only need to zero check log N many elements, i.e. the dzkp proof size. But the protocol does a very similar thing, i.e. hash the shares and send them over the network for a consistency check.